// @permaweb/aoconnect/dist/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.js
var index_exports = {};
__export(index_exports, {
  assign: () => assign,
  connect: () => connect,
  createDataItemSigner: () => createDataItemSigner2,
  createSigner: () => createSigner2,
  dryrun: () => dryrun,
  message: () => message,
  monitor: () => monitor,
  result: () => result,
  results: () => results,
  serializeCron: () => serializeCron,
  spawn: () => spawn,
  unmonitor: () => unmonitor
});
module.exports = __toCommonJS(index_exports);

// src/index.common.js
var import_ao_scheduler_utils = require("@permaweb/ao-scheduler-utils");

// node_modules/hyper-async/dist/index.js
var Async = (fork) => ({
  fork,
  toPromise: () => new Promise((resolve, reject) => fork(reject, resolve)),
  map: (fn) => Async((rej, res) => fork(rej, (x) => res(fn(x)))),
  bimap: (f, g) => Async(
    (rej, res) => fork(
      (x) => rej(f(x)),
      (x) => res(g(x))
    )
  ),
  chain: (fn) => Async((rej, res) => fork(rej, (x) => fn(x).fork(rej, res))),
  bichain: (f, g) => Async(
    (rej, res) => fork(
      (x) => f(x).fork(rej, res),
      (x) => g(x).fork(rej, res)
    )
  ),
  fold: (f, g) => Async(
    (rej, res) => fork(
      (x) => f(x).fork(rej, res),
      (x) => g(x).fork(rej, res)
    )
  )
});
var of = (x) => Async((rej, res) => res(x));
var Resolved = (x) => Async((rej, res) => res(x));
var Rejected = (x) => Async((rej, res) => rej(x));
var fromPromise = (f) => (...args) => Async(
  (rej, res) => f(...args).then(res).catch(rej)
);

// src/client/signer.js
var import_buffer3 = require("buffer/index.js");
var import_base64url4 = __toESM(require("base64url"), 1);
var import_http_message_signatures = require("http-message-signatures");
var import_structured_headers = require("structured-headers");

// src/lib/data-item.js
var import_buffer = require("buffer/index.js");
var import_base64url = __toESM(require("base64url"), 1);
var ArBundles = __toESM(require("@dha-team/arbundles"), 1);
if (!globalThis.Buffer) globalThis.Buffer = import_buffer.Buffer;
var pkg = ArBundles.default ? ArBundles.default : ArBundles;
var { createData, DataItem, SIG_CONFIG } = pkg;
function createDataItemBytes(data, signer, opts) {
  const signerMeta = SIG_CONFIG[signer.type];
  if (!signerMeta) throw new Error(`Metadata for signature type ${signer.type} not found`);
  signerMeta.signatureType = signer.type;
  signerMeta.ownerLength = signerMeta.pubLength;
  signerMeta.signatureLength = signerMeta.sigLength;
  signerMeta.publicKey = signer.publicKey;
  const dataItem = createData(data, signerMeta, opts);
  return dataItem.getRaw();
}
function getSignatureData(dataItemBytes) {
  const dataItem = new DataItem(dataItemBytes);
  return dataItem.getSignatureData();
}
function verify(dataItemBytes) {
  return DataItem.verify(dataItemBytes);
}

// src/client/hb.js
var import_base64url3 = __toESM(require("base64url"), 1);

// src/lib/utils.js
var import_ramda = require("ramda");
var import_zod = require("zod");
var joinUrl = ({ url, path: path2 }) => {
  if (!path2) return url;
  if (path2.startsWith("/")) return joinUrl({ url, path: path2.slice(1) });
  url = new URL(url);
  url.pathname += path2;
  return url.toString();
};
function parseTags(rawTags) {
  return (0, import_ramda.pipe)(
    (0, import_ramda.defaultTo)([]),
    (0, import_ramda.reduce)(
      (map3, tag) => (0, import_ramda.pipe)(
        // [value, value, ...] || []
        (0, import_ramda.propOr)([], tag.name),
        // [value]
        (0, import_ramda.append)(tag.value),
        // { [name]: [value, value, ...] }
        (0, import_ramda.assoc)(tag.name, import_ramda.__, map3)
      )(map3),
      {}
    ),
    /**
    * If the field is only a singly list, then extract the one value.
    *
    * Otherwise, keep the value as a list.
    */
    (0, import_ramda.map)((values) => values.length > 1 ? values : values[0])
  )(rawTags);
}
function eqOrIncludes(val) {
  return (0, import_ramda.cond)([
    [(0, import_ramda.is)(String), (0, import_ramda.equals)(val)],
    [(0, import_ramda.is)(Array), (0, import_ramda.includes)(val)],
    [import_ramda.T, import_ramda.F]
  ]);
}
function errFrom(err) {
  let e;
  if ((0, import_ramda.is)(import_zod.ZodError, err)) {
    e = new Error(mapZodErr(err));
    e.stack += err.stack;
  } else if ((0, import_ramda.is)(Error, err)) {
    e = err;
  } else if ((0, import_ramda.has)("message", err)) {
    e = new Error(err.message);
  } else if ((0, import_ramda.is)(String, err)) {
    e = new Error(err);
  } else {
    e = new Error("An error occurred");
  }
  return e;
}
function mapZodErr(zodErr) {
  return (0, import_ramda.pipe)(
    (zodErr2) => (
      /**
       * Take a ZodError and flatten it's issues into a single depth array
       */
      function gatherZodIssues(zodErr3, status, contextCode) {
        return (0, import_ramda.reduce)(
          (issues, issue) => (0, import_ramda.pipe)(
            (0, import_ramda.cond)([
              /**
               * These issue codes indicate nested ZodErrors, so we resursively gather those
               * See https://github.com/colinhacks/zod/blob/HEAD/ERROR_HANDLING.md#zodissuecode
               */
              [
                (0, import_ramda.equals)(import_zod.ZodIssueCode.invalid_arguments),
                () => gatherZodIssues(issue.argumentsError, 422, "Invalid Arguments")
              ],
              [
                (0, import_ramda.equals)(import_zod.ZodIssueCode.invalid_return_type),
                () => gatherZodIssues(issue.returnTypeError, 500, "Invalid Return")
              ],
              [
                (0, import_ramda.equals)(import_zod.ZodIssueCode.invalid_union),
                // An array of ZodErrors, so map over and flatten them all
                () => (0, import_ramda.chain)((i) => gatherZodIssues(i, 400, "Invalid Union"), issue.unionErrors)
              ],
              [import_ramda.T, () => [{ ...issue, status, contextCode }]]
            ]),
            (0, import_ramda.concat)(issues)
          )(issue.code),
          [],
          zodErr3.issues
        );
      }(zodErr2, 400, "")
    ),
    /**
     * combine all zod issues into a list of { message, status }
     * summaries of each issue
     */
    (zodIssues) => (0, import_ramda.reduce)(
      (acc, zodIssue) => {
        const { message: message2, path: _path, contextCode: _contextCode } = zodIssue;
        const path2 = _path[1] || _path[0];
        const contextCode = _contextCode ? `${_contextCode} ` : "";
        acc.push(`${contextCode}'${path2}': ${message2}.`);
        return acc;
      },
      [],
      zodIssues
    ),
    (0, import_ramda.join)(" | ")
  )(zodErr);
}

// src/client/hb-encode.js
var import_base64url2 = __toESM(require("base64url"), 1);
var import_buffer2 = require("buffer/index.js");
if (!globalThis.Buffer) globalThis.Buffer = import_buffer2.Buffer;
var MAX_HEADER_LENGTH = 4096;
async function sha256(data) {
  return crypto.subtle.digest("SHA-256", data);
}
function isBytes(value) {
  return value instanceof ArrayBuffer || ArrayBuffer.isView(value);
}
function isPojo(value) {
  return !isBytes(value) && typeof value === "object" && value !== null;
}
function hbEncodeValue(value) {
  if (isBytes(value)) {
    if (value.byteLength === 0) return hbEncodeValue("");
    return [void 0, value];
  }
  if (typeof value === "string") {
    if (value.length === 0) return [void 0, "empty-binary"];
    return [void 0, value];
  }
  if (Array.isArray(value) && value.length === 0) {
    return ["empty-list", void 0];
  }
  if (typeof value === "number") {
    if (!Number.isInteger(value)) return ["float", `${value}`];
    return ["integer", `${value}`];
  }
  if (typeof value === "symbol") {
    return ["atom", value.description];
  }
  throw new Error(`Cannot encode value: ${value.toString()}`);
}
function hbEncodeLift(obj, parent = "", top = {}) {
  const [flattened, types] = Object.entries(obj).reduce((acc, [key, value]) => {
    const flatK = (parent ? `${parent}/${key}` : key).toLowerCase();
    if (value == null) return acc;
    if (isPojo(value)) {
      hbEncodeLift(value, flatK, top);
      return acc;
    }
    const [type, encoded] = hbEncodeValue(value);
    if (encoded) {
      if (Buffer.from(encoded).byteLength > MAX_HEADER_LENGTH) {
        top[flatK] = encoded;
      } else acc[0][key] = encoded;
    }
    if (type) acc[1][key] = type;
    return acc;
  }, [{}, {}]);
  if (Object.keys(flattened).length === 0) return top;
  if (Object.keys(types).length > 0) {
    const aoTypes = Object.entries(types).map(([key, value]) => `${key.toLowerCase()}=${value}`).join(",");
    if (Buffer.from(aoTypes).byteLength > MAX_HEADER_LENGTH) {
      const flatK = parent ? `${parent}/ao-types` : "ao-types";
      top[flatK] = aoTypes;
    } else flattened["ao-types"] = aoTypes;
  }
  if (parent) top[parent] = flattened;
  else Object.assign(top, flattened);
  return top;
}
function encodePart(name, { headers, body }) {
  const parts = Object.entries(Object.fromEntries(headers)).reduce((acc, [name2, value]) => {
    acc.push(`${name2}: `, value, "\r\n");
    return acc;
  }, [`content-disposition: form-data;name="${name}"\r
`]);
  if (body) parts.push("\r\n", body);
  return new Blob(parts);
}
async function encode(obj = {}) {
  if (Object.keys(obj) === 0) return;
  if (obj.type !== "Process") {
    obj = Object.keys(obj).reduce((acc, h2) => {
      if (h2 !== "device") {
        acc[h2] = obj[h2];
      }
      return acc;
    }, {});
  }
  const flattened = hbEncodeLift(obj);
  const bodyKeys = [];
  const headerKeys = [];
  await Promise.all(
    Object.keys(flattened).map(async (key) => {
      const value = flattened[key];
      if (isPojo(value)) {
        const subPart = await encode(value);
        if (!subPart) return;
        bodyKeys.push(key);
        flattened[key] = encodePart(key, subPart);
        return;
      }
      if (key.includes("/") || Buffer.from(value).byteLength > MAX_HEADER_LENGTH) {
        bodyKeys.push(key);
        flattened[key] = new Blob([
          `content-disposition: form-data;name="${key}"\r
\r
`,
          value
        ]);
        return;
      }
      headerKeys.push(key);
      flattened[key] = value;
    })
  );
  const h = new Headers();
  headerKeys.forEach((key) => h.append(key, flattened[key]));
  let body;
  if (bodyKeys.length) {
    const bodyParts = await Promise.all(
      bodyKeys.map((name) => flattened[name].arrayBuffer())
    );
    const base = new Blob(
      bodyParts.flatMap((p, i, arr) => i < arr.length - 1 ? [p, "\r\n"] : [p])
    );
    const hash = await sha256(await base.arrayBuffer());
    const boundary = import_base64url2.default.encode(Buffer.from(hash));
    const blobParts = bodyParts.flatMap((p) => [`--${boundary}\r
`, p, "\r\n"]);
    blobParts.push(`--${boundary}--`);
    body = new Blob(blobParts);
    const contentDigest = await sha256(await body.arrayBuffer());
    const base64 = import_base64url2.default.toBase64(import_base64url2.default.encode(contentDigest));
    h.set("Content-Type", `multipart/form-data; boundary="${boundary}"`);
    h.append("Content-Digest", `sha-256=:${base64}:`);
  }
  return { headers: h, body };
}

// src/client/hb.js
async function encodeDataItem({ processId: processId2, data, tags, anchor }) {
  const obj = {};
  if (processId2) obj.target = processId2;
  if (anchor) obj.anchor = anchor;
  if (tags) tags.forEach((t) => {
    obj[t.name.toLowerCase()] = t.value;
  });
  obj.variant = "ao.N.1";
  obj.data = data;
  const res = await encode(obj);
  if (!res) return { headers: new Headers(), body: void 0 };
  return res;
}
function toSigBaseArgs({ url, method, headers, includePath = false }) {
  headers = new Headers(headers);
  return {
    /**
     * Always sign all headers, and the path,
     * and that there is a deterministic signature
     * component ordering
     *
     * TODO: removing path from signing, for now.
     */
    fields: [
      ...headers.keys(),
      ...includePath ? ["@path"] : []
    ].sort(),
    request: { url, method, headers: { ...Object.fromEntries(headers) } }
  };
}
function httpSigName(address) {
  const decoded = import_base64url3.default.toBuffer(address);
  const hexString = [...decoded.subarray(1, 9)].map((byte) => byte.toString(16).padStart(2, "0")).join("");
  return `http-sig-${hexString}`;
}
function requestWith({ fetch: fetch2, logger: _logger, HB_URL, signer }) {
  const logger = _logger.child("request");
  return (fields) => {
    const { path: path2, method, ...restFields } = fields;
    return of({ path: path2, method, fields: restFields }).chain(fromPromise(
      ({ path: path3, method: method2, fields: fields2 }) => {
        return encode(fields2).then(({ headers, body }) => ({
          path: path3,
          method: method2,
          headers,
          body
        }));
      }
    )).chain(fromPromise(
      async ({ path: path3, method: method2, headers, body }) => toHttpSigner(signer)(toSigBaseArgs({
        url: joinUrl({ url: HB_URL, path: path3 }),
        method: method2,
        headers
        // this does not work with hyperbeam
        // includePath: true
      })).then((req) => ({ ...req, body }))
    )).map(logger.tap("Sending HTTP signed message to HB: %o")).chain(
      (request) => of(request).chain(fromPromise(
        ({ url, method: method2, headers, body }) => {
          return fetch2(url, { method: method2, headers, body, redirect: "follow" }).then(async (res) => {
            if (res.status >= 300) return res;
            const contentType = res.headers.get("content-type");
            if (contentType && contentType.includes("multipart/form-data")) {
              return res;
            } else if (contentType && contentType.includes("application/json")) {
              const body2 = await res.json();
              return { headers: res.headers, body: body2 };
            } else {
              const body2 = await res.text();
              return { headers: res.headers, body: body2 };
            }
          });
        }
      ))
    ).toPromise();
  };
}
function deployProcessWith({ fetch: fetch2, logger: _logger, HB_URL, signer }) {
  const logger = _logger.child("deployProcess");
  return (args) => {
    return of(args).chain(fromPromise(
      ({ processId: processId2, data, tags }) => encodeDataItem({ processId: processId2, data, tags })
    )).chain(fromPromise(({ headers, body }) => {
      return toHttpSigner(signer)(toSigBaseArgs({
        url: `${HB_URL}/schedule`,
        method: "POST",
        headers
      })).then((req) => ({ ...req, body }));
    })).map(logger.tap("Sending HTTP signed message to HB MU: %o")).chain(
      (request) => of(request).chain(fromPromise(
        ({ url, method, headers, body }) => fetch2(url, { method, headers, body, redirect: "follow" })
      )).bichain(
        (err) => Rejected(err),
        fromPromise(async (res) => {
          if (res.ok) return res.headers.get("process");
          throw new Error(`${res.status}: ${await res.text()}`);
        })
      ).bimap(
        logger.tap("Error encountered when deploying process via HB MU"),
        logger.tap("Successfully deployed process via HB MU")
      ).map((process2) => ({ processId: process2 }))
    ).toPromise();
  };
}
function deployMessageWith({ fetch: fetch2, logger: _logger, HB_URL, signer }) {
  const logger = _logger.child("deployMessage");
  return (args) => {
    return of(args).chain(fromPromise(
      ({ processId: processId2, data, tags, anchor }) => encodeDataItem({ processId: processId2, data, tags, anchor })
    )).chain(fromPromise(({ headers, body }) => {
      return toHttpSigner(signer)(toSigBaseArgs({
        url: `${HB_URL}/${args.processId}/schedule`,
        method: "POST",
        headers
      })).then((req) => ({ ...req, body }));
    })).map(logger.tap("Sending HTTP signed message to HB MU: %o")).chain(
      (request) => of(request).chain(fromPromise(
        ({ url, method, headers, body }) => fetch2(url, { method, headers, body, redirect: "follow" })
      )).bichain(
        (err) => Rejected(err),
        fromPromise(async (res) => {
          if (res.ok) {
            return {
              slot: res.headers.get("slot"),
              processId: args.processId
            };
          }
          throw new Error(`${res.status}: ${await res.text()}`);
        })
      ).map(logger.tap("Received slot from HB MU: %o")).bichain(
        (err) => Rejected(err),
        fromPromise(async ({ slot, processId: processId2 }) => {
          const { headers, body } = await encodeDataItem({ processId: processId2 });
          return toHttpSigner(signer)(toSigBaseArgs({
            url: `${HB_URL}/${processId2}/push&slot+integer=${slot}`,
            method: "POST",
            headers
          })).then((req) => ({ ...req, body })).then((req) => fetch2(req.url, { method: req.method, headers: req.headers, body: req.body, redirect: "follow" })).then(() => slot);
        })
      ).bimap(
        logger.tap("Error encountered when writing message via HB MU"),
        logger.tap("Successfully wrote message via HB MU")
      ).map((slot) => ({ messageId: slot }))
    ).toPromise();
  };
}
function loadResultWith({ fetch: fetch2, logger: _logger, HB_URL, signer }) {
  const logger = _logger.child("loadResult");
  return (args) => {
    return of(args).chain(fromPromise(async ({ id, processId: processId2 }) => {
      const { headers, body } = await encodeDataItem({ processId: processId2 });
      headers.append("slot+integer", id);
      headers.append("accept", "application/json");
      return toHttpSigner(signer)(toSigBaseArgs({
        url: `${HB_URL}/${processId2}/compute&slot+integer=${id}/results/json`,
        method: "POST",
        headers
      })).then((req) => ({ ...req, body }));
    })).map(logger.tap("fetching message result from HB CU: %o")).chain(
      (request) => of(request).chain(fromPromise(
        ({ url, method, headers, body }) => fetch2(url, { method, headers, body, redirect: "follow" })
      )).bichain(
        (err) => Rejected(err),
        fromPromise(async (res) => {
          if (res.ok) return res.json();
          throw new Error(`${res.status}: ${await res.text()}`);
        })
      ).bimap(
        logger.tap("Error encountered when loading result via HB CU"),
        logger.tap("Successfully loading result via HB CU")
      )
    ).toPromise();
  };
}
function queryResultsWith({ fetch: fetch2, HB_URL, logger, signer }) {
  return ({ process: process2, from, to, sort, limit }) => {
    return of().chain(fromPromise(async () => {
      const { headers, body } = await encodeDataItem({ processId });
      headers.append("accept", "application/json");
      return toHttpSigner(signer)(toSigBaseArgs({
        url: `${HB_URL}/${processId}/state/now`,
        method: "GET",
        headers
      })).then((req) => ({ ...req, body }));
    })).chain(
      (request) => of(request).chain(fromPromise(
        ({ url, method, headers, body }) => fetch2(url, { method, headers, body, redirect: "follow" })
      )).bichain(
        (err) => Rejected(err),
        fromPromise(async (res) => {
          if (res.ok) return res.json();
          throw new Error(`${res.status}: ${await res.text()}`);
        })
      ).bimap(
        logger.tap("Error encountered when loading result via HB CU"),
        logger.tap("Successfully loading result via HB CU")
      )
    ).toPromise();
  };
}
var InsufficientFunds = class extends Error {
  name = "InsufficientFunds";
};
var RedirectRequested = class extends Error {
  name = "RedirectRequested";
};
function relayerWith({ fetch: fetch2, logger, HB_URL, signer }) {
  return async (url, options) => {
    const parsed = new URL(url);
    options.headers = new Headers(options.headers);
    const relayUrl = parsed.href;
    const relayParams = new URLSearchParams({ "relay-path": relayUrl });
    const hb = `${HB_URL}/~relay@1.0/call?${relayParams.toString()}`;
    logger('Relaying "%s" through HyperBEAM Node "%s"', parsed.href, hb);
    const { headers: signedHeaders } = await toHttpSigner(signer)({
      fields: [
        ...options.headers.keys(),
        "relay-path"
      ].sort(),
      request: {
        url,
        ...options,
        headers: { ...Object.fromEntries(options.headers), "relay-path": relayUrl }
      }
    });
    return fetch2(hb, { ...options, headers: signedHeaders }).then((res) => {
      if (res.status === 400) {
        const err = new InsufficientFunds("Insufficient Funds for request!");
        err.price = res.headers.get("price");
        throw err;
      }
      if (res.status === 422) {
        const err = new RedirectRequested("Redirect with new format!");
        err.contentEncoding = res.headers.get("content-encoding");
        err.device = res.headers.get("accept-device");
        err.location = res.headers.get("location");
        throw err;
      }
      return res;
    });
  };
}

// src/client/signer.js
var { augmentHeaders, createSignatureBase, createSigningParameters, formatSignatureBase } = import_http_message_signatures.httpbis;
if (!globalThis.Buffer) globalThis.Buffer = import_buffer3.Buffer;
var toView = (value) => {
  if (ArrayBuffer.isView(value)) value = Buffer.from(value.buffer, value.byteOffset, value.byteLength);
  else if (typeof value === "string") value = import_base64url4.default.toBuffer(value);
  else throw new Error("Unexpected type. Value must be one of Uint8Array, ArrayBuffer, or base64url-encoded string");
  return value;
};
var DATAITEM_SIGNER_KIND = "ans104";
var HTTP_SIGNER_KIND = "httpsig";
var toDataItemSigner = (signer) => {
  return async ({ data, tags, target, anchor }) => {
    let resolveUnsigned;
    let createCalled;
    const dataToSign = new Promise((resolve) => {
      resolveUnsigned = resolve;
    });
    const create = async (injected) => {
      createCalled = true;
      if (injected.passthrough) return { data, tags, target, anchor };
      const { publicKey, type = 1, alg = "rsa-v1_5-sha256" } = injected;
      const unsigned = createDataItemBytes(
        data,
        { type, publicKey: toView(publicKey) },
        { target, tags, anchor }
      );
      resolveUnsigned(unsigned);
      const deepHash = await getSignatureData(unsigned);
      return deepHash;
    };
    return signer(create, DATAITEM_SIGNER_KIND).then((res) => {
      if (!createCalled) {
        throw new Error("create() must be invoked in order to construct the data to sign");
      }
      if (typeof res === "object" && res.id && res.raw) return res;
      if (!res.signature || !res.signature) {
        throw new Error("signer must return its signature and address");
      }
      const { signature } = res;
      return dataToSign.then((unsigned) => {
        return Promise.resolve(signature).then(toView).then(async (rawSig) => {
          const signedBytes = unsigned;
          signedBytes.set(rawSig, 2);
          const isValid = await verify(signedBytes);
          if (!isValid) throw new Error("Data Item signature is not valid");
          return {
            /**
             * A data item's ID is the base64url encoded
             * SHA-256 of the signature
             */
            id: await crypto.subtle.digest("SHA-256", rawSig).then((raw) => import_base64url4.default.encode(raw)),
            raw: signedBytes
          };
        });
      });
    });
  };
};
var toHttpSigner = (signer) => {
  const params = ["alg", "keyid"].sort();
  return async ({ request, fields }) => {
    let resolveUnsigned;
    let createCalled;
    const httpSig = {};
    const dataToSign = new Promise((resolve) => {
      resolveUnsigned = resolve;
    }).then();
    const create = (injected) => {
      createCalled = true;
      let { publicKey, type = 1, alg = "rsa-pss-sha512" } = injected;
      publicKey = toView(publicKey);
      const signingParameters = createSigningParameters({
        params,
        paramValues: {
          keyid: import_base64url4.default.encode(publicKey),
          alg
        }
      });
      const signatureBase = createSignatureBase({ fields }, request);
      const signatureInput = (0, import_structured_headers.serializeList)([
        [signatureBase.map(([item]) => (0, import_structured_headers.parseItem)(item)), signingParameters]
      ]);
      signatureBase.push(['"@signature-params"', [signatureInput]]);
      const base = formatSignatureBase(signatureBase);
      httpSig.signatureInput = signatureInput;
      httpSig.signatureBase = base;
      const encoded = new TextEncoder().encode(base);
      resolveUnsigned(encoded);
      return encoded;
    };
    return signer(create, HTTP_SIGNER_KIND).then((res) => {
      if (!res.signature || !res.signature) {
        throw new Error("signer must return its signature and address");
      }
      const { signature, address } = res;
      if (!createCalled) {
        throw new Error("create() must be invoked in order to construct the data to sign");
      }
      return dataToSign.then(() => {
        return Promise.resolve(signature).then(toView).then((rawSig) => {
          const withSignature = augmentHeaders(
            request.headers,
            rawSig,
            httpSig.signatureInput,
            httpSigName(address)
          );
          const signedRequest = { ...request, headers: withSignature };
          return signedRequest;
        });
      });
    });
  };
};

// src/client/ao-mu.js
function deployMessageWith2({ fetch: fetch2, MU_URL: MU_URL2, logger: _logger }) {
  const logger = _logger.child("deployMessage");
  return (args) => {
    return of(args).chain(
      fromPromise(
        ({ processId: processId2, data, tags, anchor, signer }) => toDataItemSigner(signer)({ data, tags, target: processId2, anchor })
      )
    ).map(logger.tap("Successfully built and signed data item")).chain(
      (signedDataItem) => of(signedDataItem).chain(
        fromPromise(
          async (signedDataItem2) => fetch2(MU_URL2, {
            method: "POST",
            headers: {
              "Content-Type": "application/octet-stream",
              Accept: "application/json"
            },
            redirect: "follow",
            body: signedDataItem2.raw
          })
        )
      ).bichain(
        (err) => {
          if (err.name === "RedirectRequested") {
            return Rejected(err);
          } else {
            return Rejected(
              new Error(
                `Error while communicating with MU: ${JSON.stringify(
                  err
                )}`
              )
            );
          }
        },
        fromPromise(async (res) => {
          if (res.ok) return res.json();
          throw new Error(`${res.status}: ${await res.text()}`);
        })
      ).bimap(
        logger.tap("Error encountered when writing message via MU"),
        logger.tap("Successfully wrote message via MU")
      ).map((res) => ({ res, messageId: signedDataItem.id }))
    ).toPromise();
  };
}
function deployProcessWith2({ fetch: fetch2, MU_URL: MU_URL2, logger: _logger }) {
  const logger = _logger.child("deployProcess");
  return (args) => {
    return of(args).chain(
      fromPromise(
        ({ data, tags, signer }) => toDataItemSigner(signer)({ data, tags })
      )
    ).map(logger.tap("Successfully built and signed data item")).chain(
      (signedDataItem) => of(signedDataItem).chain(
        fromPromise(
          async (signedDataItem2) => fetch2(MU_URL2, {
            method: "POST",
            headers: {
              "Content-Type": "application/octet-stream",
              Accept: "application/json"
            },
            redirect: "follow",
            body: signedDataItem2.raw
          })
        )
      ).bichain(
        (err) => Rejected(
          new Error(
            `Error while communicating with MU: ${JSON.stringify(err)}`
          )
        ),
        fromPromise(async (res) => {
          if (res.ok) return res.json();
          throw new Error(`${res.status}: ${await res.text()}`);
        })
      ).bimap(
        logger.tap("Error encountered when deploying process via MU"),
        logger.tap("Successfully deployed process via MU")
      ).map((res) => ({ res, processId: signedDataItem.id }))
    ).toPromise();
  };
}
function deployMonitorWith({ fetch: fetch2, MU_URL: MU_URL2, logger: _logger }) {
  const logger = _logger.child("deployMonitor");
  return (args) => of(args).chain(
    fromPromise(
      ({ processId: processId2, data, tags, anchor, signer }) => toDataItemSigner(signer)({ data, tags, target: processId2, anchor })
    )
  ).map(logger.tap("Successfully built and signed data item")).chain(
    (signedDataItem) => of(signedDataItem).chain(
      fromPromise(
        async (signedDataItem2) => fetch2(MU_URL2 + "/monitor/" + args.processId, {
          method: "POST",
          headers: {
            "Content-Type": "application/octet-stream",
            Accept: "application/json"
          },
          redirect: "follow",
          body: signedDataItem2.raw
        })
      )
    ).bichain(
      (err) => Rejected(
        new Error(
          `Error while communicating with MU: ${JSON.stringify(err)}`
        )
      ),
      fromPromise(async (res) => {
        if (res.ok) return { ok: true };
        throw new Error(`${res.status}: ${await res.text()}`);
      })
    ).bimap(
      logger.tap("Error encountered when subscribing to process via MU"),
      logger.tap("Successfully subscribed to process via MU")
    ).map((res) => ({ res, messageId: signedDataItem.id }))
  ).toPromise();
}
function deployUnmonitorWith({ fetch: fetch2, MU_URL: MU_URL2, logger: _logger }) {
  const logger = _logger.child("deployUnmonitor");
  return (args) => of(args).chain(
    fromPromise(
      ({ processId: processId2, data, tags, anchor, signer }) => toDataItemSigner(signer)({ data, tags, target: processId2, anchor })
    )
  ).map(logger.tap("Successfully built and signed data item")).chain(
    (signedDataItem) => of(signedDataItem).chain(
      fromPromise(
        async (signedDataItem2) => fetch2(MU_URL2 + "/monitor/" + args.processId, {
          method: "DELETE",
          headers: {
            "Content-Type": "application/octet-stream",
            Accept: "application/json"
          },
          redirect: "follow",
          body: signedDataItem2.raw
        })
      )
    ).bichain(
      (err) => Rejected(
        new Error(
          `Error while communicating with MU: ${JSON.stringify(err)}`
        )
      ),
      fromPromise(async (res) => {
        if (res.ok) return { ok: true };
        throw new Error(`${res.status}: ${await res.text()}`);
      })
    ).bimap(
      logger.tap(
        "Error encountered when unsubscribing to process via MU"
      ),
      logger.tap("Successfully unsubscribed to process via MU")
    ).map((res) => ({ res, messageId: signedDataItem.id }))
  ).toPromise();
}
function deployAssignWith({ fetch: fetch2, MU_URL: MU_URL2, logger: _logger }) {
  const logger = _logger.child("deployAssign");
  return (args) => {
    return of(args).chain(
      fromPromise(
        async ({ process: process2, message: message2, baseLayer, exclude }) => fetch2(
          `${MU_URL2}?process-id=${process2}&assign=${message2}${baseLayer ? "&base-layer" : ""}${exclude ? "&exclude=" + exclude.join(",") : ""}`,
          {
            method: "POST",
            headers: {
              "Content-Type": "application/octet-stream",
              Accept: "application/json"
            }
          }
        )
      )
    ).bichain(
      (err) => Rejected(
        new Error(
          `Error while communicating with MU: ${JSON.stringify(err)}`
        )
      ),
      fromPromise(async (res) => {
        if (res.ok) return res.json();
        throw new Error(`${res.status}: ${await res.text()}`);
      })
    ).bimap(
      logger.tap("Error encountered when writing assignment via MU"),
      logger.tap("Successfully wrote assignment via MU")
    ).map((res) => ({ res, assignmentId: res.id })).toPromise();
  };
}

// src/client/ao-cu.js
function dryrunFetchWith({ fetch: fetch2, CU_URL: CU_URL2, logger }) {
  return (msg) => of(msg).map(logger.tap("posting dryrun request to CU")).chain(fromPromise((msg2) => fetch2(`${CU_URL2}/dry-run?process-id=${msg2.Target}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    redirect: "follow",
    body: JSON.stringify(msg2)
  }).then((res) => res.json()))).toPromise();
}
function loadResultWith2({ fetch: fetch2, CU_URL: CU_URL2, logger }) {
  return ({ id, processId: processId2 }) => {
    return of(`${CU_URL2}/result/${id}?process-id=${processId2}`).map(logger.tap("fetching message result from CU")).chain(fromPromise(
      async (url) => fetch2(url, {
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        redirect: "follow"
      }).then((res) => res.json())
    )).toPromise();
  };
}
function queryResultsWith2({ fetch: fetch2, CU_URL: CU_URL2, logger }) {
  return ({ process: process2, from, to, sort, limit }) => {
    const target = new URL(`${CU_URL2}/results/${process2}`);
    const params = new URLSearchParams(target.search);
    if (from) {
      params.append("from", from);
    }
    if (to) {
      params.append("to", to);
    }
    if (sort) {
      params.append("sort", sort);
    }
    if (limit) {
      params.append("limit", limit);
    }
    target.search = params;
    return of(target.toString()).map(logger.tap("fetching message result from CU")).chain(fromPromise(
      async (url) => fetch2(url, {
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        redirect: "follow"
      }).then((res) => res.json())
    )).toPromise();
  };
}

// src/client/gateway.js
var import_ramda2 = require("ramda");
var import_zod2 = require("zod");
function loadTransactionMetaWith({ fetch: fetch2, GRAPHQL_URL: GRAPHQL_URL2, logger }) {
  const GET_TRANSACTIONS_QUERY = `
    query GetTransactions ($transactionIds: [ID!]!) {
      transactions(ids: $transactionIds) {
        edges {
          node {
            owner {
              address
            }
            tags {
              name
              value
            }
            block {
              id
              height
              timestamp
            }
          }
        }
      }
    }`;
  const transactionConnectionSchema = import_zod2.z.object({
    data: import_zod2.z.object({
      transactions: import_zod2.z.object({
        edges: import_zod2.z.array(import_zod2.z.object({
          node: import_zod2.z.record(import_zod2.z.any())
        }))
      })
    })
  });
  return (id) => of(id).chain(fromPromise(
    (id2) => fetch2(GRAPHQL_URL2, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        query: GET_TRANSACTIONS_QUERY,
        variables: { transactionIds: [id2] }
      })
    }).then(async (res) => {
      if (res.ok) return res.json();
      logger('Error Encountered when querying gateway for transaction "%s"', id2);
      throw new Error(`${res.status}: ${await res.text()}`);
    }).then(transactionConnectionSchema.parse).then((0, import_ramda2.path)(["data", "transactions", "edges", "0", "node"]))
  )).toPromise();
}

// src/client/ao-su.js
var import_lru_map = __toESM(require("mnemonist/lru-map.js"), 1);
var getMessagesByRange = ({ fetch: fetch2, locate: locate2 }) => {
  return async ({ processId: processId2, from, to, limit }) => {
    const suUrl = (await locate2(processId2)).url;
    return fetch2(`${suUrl}/${processId2}?from=${from}&to=${to}&limit=${limit}`).then((res) => res.json());
  };
};
var getMessageById = ({ fetch: fetch2, locate: locate2 }) => {
  return async ({ processId: processId2, messageId }) => {
    const suUrl = (await locate2(processId2)).url;
    return fetch2(`${suUrl}/${messageId}?process-id=${processId2}`).then((res) => res.json());
  };
};
var getLastSlotWith = ({ fetch: fetch2, locate: locate2 }) => {
  return async ({ processId: processId2, since }) => {
    if (!since) {
      since = 10 * 60 * 1e3;
    }
    const from = Date.now() - since;
    const suUrl = (await locate2(processId2)).url;
    return fetch2(`${suUrl}/${processId2}?process-id=${processId2}&limit=100&from=${from}`).then((res) => res.json()).then((suResult) => {
      return suResult.edges[suResult?.edges?.length - 1]?.node?.assignment?.tags?.find((t) => t.name === "Nonce")?.value;
    });
  };
};

// src/logger.js
var import_debug = __toESM(require("debug"), 1);
var import_ramda3 = require("ramda");
var createLogger = (name = "@permaweb/aoconnect") => {
  const logger = (0, import_debug.default)(name);
  logger.child = (name2) => createLogger(`${logger.namespace}:${name2}`);
  logger.tap = (note, ...rest) => (0, import_ramda3.tap)((...args) => logger(note, ...rest, ...args));
  return logger;
};

// src/lib/messages/index.js
var import_zod3 = require("zod");
var inputSchema = import_zod3.z.object({
  processId: import_zod3.z.string(),
  from: import_zod3.z.string(),
  to: import_zod3.z.string(),
  limit: import_zod3.z.string().default("1000").optional()
}).passthrough();
function messagesWith(env) {
  const messages = env.messages;
  return (fields) => {
    fields = inputSchema.parse(fields);
    return messages(fields);
  };
}

// src/lib/message-id/index.js
var import_zod4 = require("zod");
var inputSchema2 = import_zod4.z.object({
  processId: import_zod4.z.string(),
  messageId: import_zod4.z.string()
}).passthrough();
function messageIdWith(env) {
  const getMessageId = env.getMessageId;
  return (fields) => {
    fields = inputSchema2.parse(fields);
    return getMessageId(fields);
  };
}

// src/lib/process-id/index.js
var import_zod5 = require("zod");
var inputSchema3 = import_zod5.z.object({
  processId: import_zod5.z.string(),
  since: import_zod5.z.number().optional()
}).passthrough();
function processIdWith(env) {
  const processId2 = env.processId;
  return (fields) => {
    fields = inputSchema3.parse(fields);
    return processId2(fields);
  };
}

// src/lib/request/index.js
var import_ramda5 = require("ramda");
var import_zod6 = require("zod");

// src/lib/request/transform.js
var transformToMap = (device) => (ctx) => {
  const result2 = ctx.res;
  const map3 = {};
  if (device === "relay@1.0") {
    if (typeof result2 === "string") {
      return result2;
    }
    if (result2.Output && result2.Output.data) {
      map3.Output = result2.Output;
    }
    if (result2.Messages) {
      map3.Messages = result2.Messages.map((m) => {
        const miniMap = {};
        m.Tags.forEach((t) => {
          miniMap[t.name] = t.value;
        });
        miniMap.Data = {
          text: () => Promise.resolve(m.Data),
          json: () => Promise.resolve(JSON.parse(m.Data))
          // binary: () => Promise.resolve(Buffer.from(m.Data))
        };
        miniMap.Target = m.Target;
        miniMap.Anchor = m.Anchor;
        return miniMap;
      });
    }
    return map3;
  } else {
    map3.process = ctx.process;
    map3.slot = ctx.slot;
    const res = result2;
    let body = "";
    res.headers.forEach((v, k) => {
      map3[k] = v;
    });
    if (typeof res.body === "string") {
      try {
        body = JSON.parse(res.body);
        if (body.Output) {
          map3.Output = body.Output;
        }
        if (body.Messages) {
          map3.Messages = body.Messages.map((m) => {
            const miniMap = {};
            m.Tags.forEach((t) => {
              miniMap[t.name] = t.value;
            });
            miniMap.Data = {
              text: () => Promise.resolve(m.Data),
              json: () => Promise.resolve(JSON.parse(m.Data))
              // import { Buffer as BufferShim } from 'buffer/index.js'
              // if (!globalThis.Buffer) globalThis.Buffer = BufferShim
              // binary: () => Promise.resolve(Buffer.from(m.Data))
            };
            miniMap.Target = m.Target;
            miniMap.Anchor = m.Anchor;
            return miniMap;
          });
        }
      } catch (e) {
        map3.body = { text: () => Promise.resolve(body) };
      }
    }
    return map3;
  }
};

// src/lib/request/dispatch.js
function dispatch({ request, spawn: spawn2, message: message2, result: result2, dryrun: dryrun2, signer }) {
  return function(ctx) {
    if (ctx.type === "dryrun" && ctx.dataItem) {
      const inputData = {
        process: ctx.dataItem.target,
        anchor: ctx.dataItem.anchor,
        tags: ctx.dataItem.tags,
        data: ctx.datatItem?.data ?? ""
      };
      return fromPromise(
        () => dryrun2(inputData).catch((err) => {
          if (err.message.includes("Insufficient funds")) {
            return { error: "insufficient-funds" };
          }
          throw err;
        }).then((res) => ({ res }))
      )(ctx);
    }
    if (ctx.type === "Message" && ctx.dataItem) {
      return fromPromise(
        (ctx2) => message2({
          process: ctx2.dataItem.target,
          anchor: ctx2.dataItem.anchor,
          tags: ctx2.dataItem.tags,
          data: ctx2.dataItem.data ?? "",
          signer
        }).catch((err) => {
          if (err.message.includes("Insufficient funds")) {
            return Rejected({ error: "insufficient-funds" });
          }
          return Rejected(err);
        })
      )(ctx).chain(fromPromise((msg) => result2({ process: ctx.dataItem.target, message: msg }))).map((res) => ({ res }));
    }
    if (ctx.type === "Process" && ctx.dataItem) {
      return fromPromise(
        (ctx2) => spawn2({
          tags: ctx2.dataItem.tags,
          data: ctx2.dataItem.data,
          scheduler: ctx2.dataItem.tags.find((t) => t.name === "Scheduler")?.value,
          module: ctx2.dataItem.tags.find((t) => t.name === "Module")?.value,
          signer
        }).then((res) => ({ res })).catch((err) => {
          if (err.message.includes("Insufficient funds")) {
            return { error: "insufficient-funds" };
          }
          throw err;
        })
      )(ctx).bichain((err) => {
        console.log(ctx);
        if (err.name === "RedirectRequested") {
        }
      }, of);
    }
    if (ctx.map) {
      return fromPromise((ctx2) => {
        return request(ctx2.map).then((res) => {
          return { res };
        });
      })(ctx);
    }
  };
}

// src/lib/request/format.js
var import_ramda4 = require("ramda");
var handleFormat = (mode, device) => (fields) => {
  if (mode === "mainnet" && device === "relay@1.0") {
    const dataItem = {
      target: fields.Target ?? fields.process,
      anchor: fields.Anchor ?? "",
      tags: (0, import_ramda4.keys)(
        (0, import_ramda4.omit)(
          [
            "Target",
            "Anchor",
            "Data",
            "dryrun",
            "Type",
            "Variant",
            "path",
            "method"
          ],
          fields
        )
      ).map(function(key) {
        return { name: key, value: fields[key] };
      }, fields).concat([
        { name: "Data-Protocol", value: "ao" },
        { name: "Type", value: fields.Type ?? "Message" },
        { name: "Variant", value: fields.Variant ?? "ao.N.1" }
      ]),
      data: fields?.data || ""
    };
    let _type = fields.Type ?? "Message";
    if (fields.dryrun) {
      _type = "dryrun";
    }
    return {
      type: _type,
      dataItem
    };
  }
  if (mode === "mainnet") {
    const map3 = fields;
    if (!fields.path) {
      fields.path = "/schedule";
    }
    if (fields.process && !fields.path.includes(fields.process)) {
      fields.path = `${fields.process}${fields.path}`;
    }
    return {
      type: fields.Type,
      map: map3
    };
  }
};

// src/lib/request/index.js
var inputSchema4 = import_zod6.z.object({
  path: import_zod6.z.string().min(1, { message: "path is required" }),
  method: import_zod6.z.string()
}).passthrough();
function requestWith2(env) {
  const logger = env.logger;
  const request = env.request;
  const message2 = env.message;
  const result2 = env.result;
  const dryrun2 = env.dryrun;
  const spawn2 = env.spawn;
  const signer = env.signer;
  const device = env.device;
  const method = env.method;
  const mode = env.MODE;
  const verifyInput = (args) => {
    if (args.device === "relay@1.0" && args.Type === "Process") {
      return of(
        import_zod6.z.object({
          path: import_zod6.z.string().min(1, { message: "path is required" }),
          method: import_zod6.z.string(),
          Module: import_zod6.z.string(),
          Scheduler: import_zod6.z.string()
        }).passthrough().parse(args)
      );
    }
    return of(inputSchema4.parse(args));
  };
  return (fields) => {
    const retry = (fn, request2, attempts) => fn(request2).bichain((err) => {
      if (err.name !== "RedirectRequested") return Rejected(err);
      if (attempts <= 0) return Rejected(err);
      const newRequest = {
        ...request2,
        device: err.device
      };
      return retry(fn, newRequest, attempts - 1);
    }, of);
    const operation = (ctx) => verifyInput(ctx).map(handleFormat(mode, device)).chain(dispatch({ request, spawn: spawn2, message: message2, result: result2, dryrun: dryrun2, signer })).map((_) => {
      logger(
        'Received response from message sent to path "%s"',
        fields?.path ?? "/"
      );
      return _;
    }).map(transformToMap(device)).bimap(errFrom, import_ramda5.identity);
    return retry(operation, { path: `/~${device}`, ...fields, method: fields.method ?? method }, 1).toPromise();
  };
}

// src/lib/result/index.js
var import_ramda6 = require("ramda");

// src/lib/result/verify-input.js
var import_zod7 = require("zod");
var inputSchema5 = import_zod7.z.object({
  id: import_zod7.z.string().min(1, { message: "message is required to be a message id" }),
  processId: import_zod7.z.string().min(1, { message: "process is required to be a process id" })
});
function verifyInputWith() {
  return (ctx) => {
    return of(ctx).map(inputSchema5.parse).map(() => ctx);
  };
}

// src/dal.js
var import_zod8 = require("zod");
var tagSchema = import_zod8.z.object({
  name: import_zod8.z.string(),
  value: import_zod8.z.any()
});
var dryrunResultSchema = import_zod8.z.function().args(import_zod8.z.object({
  Id: import_zod8.z.string(),
  Target: import_zod8.z.string(),
  Owner: import_zod8.z.string(),
  Anchor: import_zod8.z.string().optional(),
  Data: import_zod8.z.any().default("1234"),
  Tags: import_zod8.z.array(import_zod8.z.object({ name: import_zod8.z.string(), value: import_zod8.z.string() }))
})).returns(import_zod8.z.promise(import_zod8.z.any()));
var loadResultSchema = import_zod8.z.function().args(import_zod8.z.object({
  id: import_zod8.z.string().min(1, { message: "message id is required" }),
  processId: import_zod8.z.string().min(1, { message: "process id is required" })
})).returns(import_zod8.z.promise(import_zod8.z.any()));
var queryResultsSchema = import_zod8.z.function().args(import_zod8.z.object({
  process: import_zod8.z.string().min(1, { message: "process id is required" }),
  from: import_zod8.z.string().optional(),
  to: import_zod8.z.string().optional(),
  sort: import_zod8.z.enum(["ASC", "DESC"]).default("ASC"),
  limit: import_zod8.z.number().optional()
})).returns(import_zod8.z.promise(import_zod8.z.object({
  edges: import_zod8.z.array(import_zod8.z.object({
    cursor: import_zod8.z.string(),
    node: import_zod8.z.object({
      Output: import_zod8.z.any().optional(),
      Messages: import_zod8.z.array(import_zod8.z.any()).optional(),
      Spawns: import_zod8.z.array(import_zod8.z.any()).optional(),
      Error: import_zod8.z.any().optional()
    })
  }))
})));
var deployMessageSchema = import_zod8.z.function().args(import_zod8.z.object({
  processId: import_zod8.z.string(),
  data: import_zod8.z.any(),
  tags: import_zod8.z.array(tagSchema),
  anchor: import_zod8.z.string().optional(),
  signer: import_zod8.z.any().nullish()
})).returns(import_zod8.z.promise(
  import_zod8.z.object({
    messageId: import_zod8.z.string()
  }).passthrough()
));
var deployProcessSchema = import_zod8.z.function().args(import_zod8.z.object({
  data: import_zod8.z.any(),
  tags: import_zod8.z.array(tagSchema),
  signer: import_zod8.z.any().nullish()
})).returns(import_zod8.z.promise(
  import_zod8.z.object({
    processId: import_zod8.z.string()
  }).passthrough()
));
var deployAssignSchema = import_zod8.z.function().args(import_zod8.z.object({
  process: import_zod8.z.string(),
  message: import_zod8.z.string(),
  baseLayer: import_zod8.z.boolean().optional(),
  exclude: import_zod8.z.array(import_zod8.z.string()).optional()
})).returns(import_zod8.z.promise(
  import_zod8.z.object({
    assignmentId: import_zod8.z.string()
  }).passthrough()
));
var deployMonitorSchema = deployMessageSchema;
var loadProcessMetaSchema = import_zod8.z.function().args(import_zod8.z.object({
  suUrl: import_zod8.z.string().url(),
  processId: import_zod8.z.string()
})).returns(import_zod8.z.promise(
  import_zod8.z.object({
    tags: import_zod8.z.array(tagSchema)
  }).passthrough()
));
var locateSchedulerSchema = import_zod8.z.function().args(import_zod8.z.string()).returns(import_zod8.z.promise(
  import_zod8.z.object({
    url: import_zod8.z.string()
  })
));
var validateSchedulerSchema = import_zod8.z.function().args(import_zod8.z.string()).returns(import_zod8.z.promise(import_zod8.z.boolean()));
var loadTransactionMetaSchema = import_zod8.z.function().args(import_zod8.z.string()).returns(import_zod8.z.promise(
  import_zod8.z.object({
    tags: import_zod8.z.array(tagSchema)
  }).passthrough()
));
var signerSchema = import_zod8.z.function();

// src/lib/result/read.js
function readWith({ loadResult }) {
  loadResult = fromPromise(loadResultSchema.implement(loadResult));
  return (ctx) => {
    return of({ id: ctx.id, processId: ctx.processId }).chain(loadResult);
  };
}

// src/lib/result/index.js
function resultWith(env) {
  const verifyInput = verifyInputWith(env);
  const read = readWith(env);
  return ({ message: message2, process: process2 }) => {
    return of({ id: message2, processId: process2 }).chain(verifyInput).chain(read).map(
      env.logger.tap(
        'readResult result for message "%s": %O',
        message2
      )
    ).map((result2) => result2).bimap(errFrom, import_ramda6.identity).toPromise();
  };
}

// src/lib/message/index.js
var import_ramda8 = require("ramda");

// src/lib/message/upload-message.js
var import_zod9 = require("zod");
var import_ramda7 = require("ramda");
var import_protocol_tag_utils = require("@permaweb/protocol-tag-utils");
var aoProto = (0, import_protocol_tag_utils.proto)("ao");
var removeAoProtoByName = (0, import_ramda7.curry)(aoProto.removeAllByName);
var concatAoProto = (0, import_ramda7.curry)(aoProto.concat);
var concatUnassoc = (0, import_ramda7.curry)(aoProto.concatUnassoc);
var tagSchema2 = import_zod9.z.array(import_zod9.z.object({
  name: import_zod9.z.string(),
  value: import_zod9.z.string()
}));
function buildTagsWith() {
  return (ctx) => {
    return of(ctx.tags).map((0, import_ramda7.defaultTo)([])).map(removeAoProtoByName("Variant")).map(removeAoProtoByName("Type")).map(concatAoProto([
      { name: "Variant", value: "ao.TN.1" },
      { name: "Type", value: "Message" }
    ])).map(tagSchema2.parse).map((0, import_ramda7.assoc)("tags", import_ramda7.__, ctx));
  };
}
function buildDataWith({ logger }) {
  return (ctx) => {
    return of(ctx).chain((0, import_ramda7.ifElse)(
      (0, import_ramda7.always)(ctx.data),
      /**
       * data is provided as input, so do nothing
       */
      () => Resolved(ctx),
      /**
       * No data is provided, so replace with one space
       */
      () => Resolved(" ").map((0, import_ramda7.assoc)("data", import_ramda7.__, ctx)).map(
        (ctx2) => (0, import_ramda7.pipe)(
          (0, import_ramda7.prop)("tags"),
          concatUnassoc([{ name: "Content-Type", value: "text/plain" }]),
          (0, import_ramda7.assoc)("tags", import_ramda7.__, ctx2)
        )(ctx2)
      ).map(logger.tap('added pseudo-random string as message "data"'))
    )).map(
      (ctx2) => (0, import_ramda7.pipe)(
        (0, import_ramda7.prop)("tags"),
        concatUnassoc([{ name: "SDK", value: "aoconnect" }]),
        (0, import_ramda7.assoc)("tags", import_ramda7.__, ctx2)
      )(ctx2)
    );
  };
}
function uploadMessageWith(env) {
  const buildTags = buildTagsWith(env);
  const buildData = buildDataWith(env);
  const deployMessage = deployMessageSchema.implement(env.deployMessage);
  return (ctx) => {
    return of(ctx).chain(buildTags).chain(buildData).chain(fromPromise(
      ({ id, data, tags, anchor, signer }) => deployMessage({ processId: id, data, tags, anchor, signer: signerSchema.implement(signer || env.signer) })
    )).map((res) => (0, import_ramda7.assoc)("messageId", res.messageId, ctx));
  };
}

// src/lib/message/index.js
function messageWith(env) {
  const uploadMessage = uploadMessageWith(env);
  return ({ process: process2, data, tags, anchor, signer }) => {
    return of({ id: process2, data, tags, anchor, signer }).chain(uploadMessage).map((ctx) => ctx.messageId).bimap(errFrom, import_ramda8.identity).toPromise();
  };
}

// src/lib/spawn/index.js
var import_ramda11 = require("ramda");

// src/lib/spawn/verify-inputs.js
var import_ramda9 = require("ramda");
var checkTag = (name, pred, err) => (tags) => pred(tags[name]) ? Resolved(tags) : Rejected(`Tag '${name}': ${err}`);
function verifyModuleWith({ loadTransactionMeta, logger }) {
  loadTransactionMeta = fromPromise(loadTransactionMetaSchema.implement(loadTransactionMeta));
  return (module2) => of(module2).chain(loadTransactionMeta).map((0, import_ramda9.prop)("tags")).map(parseTags).chain(checkTag("Data-Protocol", eqOrIncludes("ao"), "value 'ao' was not found on module")).chain(checkTag("Type", eqOrIncludes("Module"), "value 'Module' was not found on module")).chain(checkTag("Module-Format", import_ramda9.isNotNil, "was not found on module")).chain(checkTag("Input-Encoding", import_ramda9.isNotNil, "was not found on module")).chain(checkTag("Output-Encoding", import_ramda9.isNotNil, "was not found on module")).bimap(
    logger.tap("Verifying module source failed: %s"),
    logger.tap("Verified module source")
  );
}
function verifySchedulerWith({ logger, validateScheduler }) {
  validateScheduler = fromPromise(validateSchedulerSchema.implement(validateScheduler));
  return (scheduler) => of(scheduler).chain(
    (scheduler2) => validateScheduler(scheduler2).chain((isValid) => isValid ? Resolved(scheduler2) : Rejected(`Valid Scheduler-Location owned by ${scheduler2} not found`))
  ).bimap(
    logger.tap("Verifying scheduler failed: %s"),
    logger.tap("Verified scheduler")
  );
}
function verifySignerWith({ logger }) {
  return (signer) => of(signer).map(logger.tap("Checking for signer")).chain((signer2) => signer2 ? Resolved(signer2) : Rejected("signer not found"));
}
function verifyInputsWith(env) {
  const logger = env.logger.child("verifyInput");
  env = { ...env, logger };
  const verifyModule = verifyModuleWith(env);
  const verifyScheduler = verifySchedulerWith(env);
  const verifySigner = verifySignerWith(env);
  return (ctx) => {
    return of(ctx).chain((ctx2) => verifyModule(ctx2.module).map(() => ctx2)).bimap(
      logger.tap("Error when verify input: %s"),
      logger.tap("Successfully verified inputs")
    );
  };
}

// src/lib/spawn/upload-process.js
var import_zod10 = require("zod");
var import_ramda10 = require("ramda");
var import_protocol_tag_utils2 = require("@permaweb/protocol-tag-utils");
var aoProto2 = (0, import_protocol_tag_utils2.proto)("ao");
var removeAoProtoByName2 = (0, import_ramda10.curry)(aoProto2.removeAllByName);
var concatAoProto2 = (0, import_ramda10.curry)(aoProto2.concat);
var concatUnassoc2 = (0, import_ramda10.curry)(aoProto2.concatUnassoc);
var tagsSchema = import_zod10.z.array(tagSchema);
function buildTagsWith2() {
  return (ctx) => {
    return of(ctx).map((0, import_ramda10.prop)("tags")).map((0, import_ramda10.defaultTo)([])).map(removeAoProtoByName2("Variant")).map(removeAoProtoByName2("Type")).map(removeAoProtoByName2("Module")).map(removeAoProtoByName2("Scheduler")).map(concatAoProto2([
      { name: "Variant", value: "ao.TN.1" },
      { name: "Type", value: "Process" },
      { name: "Module", value: ctx.module },
      { name: "Scheduler", value: ctx.scheduler }
    ])).map(tagsSchema.parse).map((0, import_ramda10.assoc)("tags", import_ramda10.__, ctx));
  };
}
function buildDataWith2({ logger }) {
  return (ctx) => {
    return of(ctx).chain((0, import_ramda10.ifElse)(
      (0, import_ramda10.always)(ctx.data),
      /**
       * data is provided as input, so do nothing
       */
      () => Resolved(ctx),
      /**
       * No data is provided, so replace with one space
       */
      () => Resolved(" ").map((0, import_ramda10.assoc)("data", import_ramda10.__, ctx)).map(
        (ctx2) => (0, import_ramda10.pipe)(
          (0, import_ramda10.prop)("tags"),
          concatUnassoc2([{ name: "Content-Type", value: "text/plain" }]),
          (0, import_ramda10.assoc)("tags", import_ramda10.__, ctx2)
        )(ctx2)
      ).map(logger.tap('added pseudo-random string as process "data"'))
    )).map(
      (ctx2) => (0, import_ramda10.pipe)(
        (0, import_ramda10.prop)("tags"),
        concatUnassoc2([{ name: "SDK", value: "aoconnect" }]),
        (0, import_ramda10.assoc)("tags", import_ramda10.__, ctx2)
      )(ctx2)
    );
  };
}
function uploadProcessWith(env) {
  const logger = env.logger.child("uploadProcess");
  env = { ...env, logger };
  const buildTags = buildTagsWith2(env);
  const buildData = buildDataWith2(env);
  const deployProcess = deployProcessSchema.implement(env.deployProcess);
  return (ctx) => {
    return of(ctx).chain(buildTags).chain(buildData).chain(fromPromise(
      ({ data, tags, signer }) => deployProcess({ data, tags, signer: signerSchema.implement(signer || env.signer) })
    )).map((res) => (0, import_ramda10.assoc)("processId", res.processId, ctx));
  };
}

// src/lib/spawn/index.js
function spawnWith(env) {
  const verifyInputs = verifyInputsWith(env);
  const uploadProcess = uploadProcessWith(env);
  return ({ module: module2, scheduler, signer, tags, data }) => {
    return of({ module: module2, scheduler, signer, tags, data }).chain(verifyInputs).chain(uploadProcess).map((ctx) => ctx.processId).bimap(errFrom, import_ramda11.identity).toPromise();
  };
}

// src/lib/monitor/index.js
var import_ramda13 = require("ramda");

// src/lib/monitor/upload-monitor.js
var import_ramda12 = require("ramda");
function uploadMonitorWith(env) {
  const deployMonitor = deployMonitorSchema.implement(env.deployMonitor);
  return (ctx) => {
    return of(ctx).chain(fromPromise(
      ({ id, signer }) => deployMonitor({
        processId: id,
        signer: signerSchema.implement(signer || env.signer),
        /**
         * No tags or data can be provided right now,
         *
         * so just set data to single space and set tags to an empty array
         */
        data: " ",
        tags: []
      })
    )).map((res) => (0, import_ramda12.assoc)("monitorId", res.messageId, ctx));
  };
}

// src/lib/monitor/index.js
function monitorWith(env) {
  const uploadMonitor = uploadMonitorWith(env);
  return ({ process: process2, signer }) => of({ id: process2, signer }).chain(uploadMonitor).map((ctx) => ctx.monitorId).bimap(errFrom, import_ramda13.identity).toPromise();
}

// src/lib/unmonitor/index.js
var import_ramda15 = require("ramda");

// src/lib/unmonitor/upload-unmonitor.js
var import_ramda14 = require("ramda");
function uploadUnmonitorWith(env) {
  const deployUnmonitor = deployMonitorSchema.implement(env.deployUnmonitor);
  return (ctx) => {
    return of(ctx).chain(fromPromise(
      ({ id, signer }) => deployUnmonitor({
        processId: id,
        signer: signerSchema.implement(signer || env.signer),
        /**
         * No tags or data can be provided right now,
         *
         * so just set data to single space and set tags to an empty array
         */
        data: " ",
        tags: []
      })
    )).map((res) => (0, import_ramda14.assoc)("monitorId", res.messageId, ctx));
  };
}

// src/lib/unmonitor/index.js
function unmonitorWith(env) {
  const uploadUnmonitor = uploadUnmonitorWith(env);
  return ({ process: process2, signer }) => of({ id: process2, signer }).chain(uploadUnmonitor).map((ctx) => ctx.monitorId).bimap(errFrom, import_ramda15.identity).toPromise();
}

// src/lib/results/index.js
var import_ramda16 = require("ramda");

// src/lib/results/verify-input.js
var import_zod11 = require("zod");
var inputSchema6 = import_zod11.z.object({
  process: import_zod11.z.string().min(1, { message: "process identifier is required" }),
  from: import_zod11.z.string().optional(),
  to: import_zod11.z.string().optional(),
  sort: import_zod11.z.enum(["ASC", "DESC"]).default("ASC"),
  limit: import_zod11.z.number().optional()
});
function verifyInputWith2() {
  return (ctx) => {
    return of(ctx).map(inputSchema6.parse).map(() => ctx);
  };
}

// src/lib/results/query.js
function queryWith({ queryResults }) {
  queryResults = fromPromise(queryResultsSchema.implement(queryResults));
  return (ctx) => {
    return of({ process: ctx.process, from: ctx.from, to: ctx.to, sort: ctx.sort, limit: ctx.limit }).chain(queryResults);
  };
}

// src/lib/results/index.js
function resultsWith(env) {
  const verifyInput = verifyInputWith2(env);
  const query = queryWith(env);
  return ({ process: process2, from, to, sort, limit }) => {
    return of({ process: process2, from, to, sort, limit }).chain(verifyInput).chain(query).map(
      env.logger.tap(
        'readResults result for message "%s": %O',
        process2
      )
    ).map((result2) => result2).bimap(errFrom, import_ramda16.identity).toPromise();
  };
}

// src/lib/dryrun/verify-input.js
var import_zod12 = require("zod");
var inputSchema7 = import_zod12.z.object({
  Id: import_zod12.z.string(),
  Target: import_zod12.z.string(),
  Owner: import_zod12.z.string(),
  Anchor: import_zod12.z.string().optional(),
  Data: import_zod12.z.any().default("1234"),
  Tags: import_zod12.z.array(import_zod12.z.object({ name: import_zod12.z.string(), value: import_zod12.z.string() }))
});
function verifyInputWith3() {
  return (msg) => {
    return of(msg).map(inputSchema7.parse).map((m) => {
      m.Tags = m.Tags.concat([
        { name: "Data-Protocol", value: "ao" },
        { name: "Type", value: "Message" },
        { name: "Variant", value: "ao.TN.1" }
      ]);
      return m;
    });
  };
}

// src/lib/dryrun/run.js
function runWith({ dryrunFetch }) {
  return fromPromise(dryrunResultSchema.implement(dryrunFetch));
}

// src/lib/dryrun/index.js
function dryrunWith(env) {
  const verifyInput = verifyInputWith3(env);
  const dryrun2 = runWith(env);
  return (msg) => of(msg).map(convert).chain(verifyInput).chain(dryrun2).toPromise();
}
function convert({ process: process2, data, tags, anchor, ...rest }) {
  return {
    Id: "1234",
    Owner: "1234",
    ...rest,
    Target: process2,
    Data: data || "1234",
    Tags: tags || [],
    Anchor: anchor || "0"
  };
}

// src/lib/assign/index.js
var import_ramda18 = require("ramda");

// src/lib/assign/send-assign.js
var import_ramda17 = require("ramda");
function sendAssignWith(env) {
  const deployAssign = deployAssignSchema.implement(env.deployAssign);
  return (ctx) => {
    return of(ctx).chain(fromPromise(
      ({ process: process2, message: message2, baseLayer, exclude }) => deployAssign({ process: process2, message: message2, baseLayer, exclude })
    )).map((res) => (0, import_ramda17.assoc)("assignmentId", res.assignmentId, ctx));
  };
}

// src/lib/assign/index.js
function assignWith(env) {
  const sendAssign = sendAssignWith(env);
  return ({ process: process2, message: message2, baseLayer, exclude }) => {
    return of({ process: process2, message: message2, baseLayer, exclude }).chain(sendAssign).map((ctx) => ctx.assignmentId).bimap(errFrom, import_ramda18.identity).toPromise();
  };
}

// src/lib/payments/index.js
function getOperator(env) {
  return () => of({ URL: env.URL, fetch: env.fetch }).chain(getAddress).toPromise();
}
function getNodeBalance(env) {
  return () => of().chain(requestBalance(env.request)).toPromise();
}
function requestBalance(request) {
  return fromPromise(() => request({
    method: "GET",
    path: "/~simple-pay@1.0/balance"
  }).then((res) => res.body));
}
function getAddress({ URL: URL2, fetch: fetch2 }) {
  return fromPromise(
    (nodeUrl) => fetch2(`${nodeUrl}/~meta@1.0/info/address`).then((res) => res.text()).catch((e) => "N/A")
  )(URL2);
}

// src/lib/serializeCron/index.js
var import_ramda19 = require("ramda");
function serializeCron(cron) {
  function parseInterval(interval2 = "") {
    if (typeof interval2 !== "string") throw new Error("Encountered Error serializing cron: invalid interval");
    const [value, unit] = interval2.split("-").map((s) => s.trim());
    if (!value || !unit) throw new Error("Encountered Error serializing cron: invalid interval");
    if (!parseInt(value) || parseInt(value) < 0) throw new Error("Encountered Error serializing cron: invalid interval value");
    const singularRegex = /^(millisecond|second|minute|hour|day|month|year|block)$/;
    const pluralRegex = /^(milliseconds|seconds|minutes|hours|days|months|years|blocks)$/;
    const unitSingularMatch = unit.match(singularRegex);
    const unitPluralMatch = unit.match(pluralRegex);
    if (parseInt(value) > 1 && !unitPluralMatch || parseInt(value) === 1 && !unitSingularMatch) throw new Error("Encountered Error serializing cron: invalid interval type");
    return `${value}-${unit}`;
  }
  function parseTags2(tags2 = []) {
    return (0, import_ramda19.map)((tag) => {
      if (!tag.name || !tag.value) throw new Error("Encountered Error serializing cron: invalid tag structure");
      if (typeof tag.name !== "string" || typeof tag.value !== "string") throw new Error("Encountered Error serializing cron: invalid interval tag types");
      return { name: `Cron-Tag-${tag.name}`, value: tag.value };
    }, tags2);
  }
  const interval = parseInterval(cron.interval);
  const tags = parseTags2(cron.tags);
  return [{ name: "Cron-Interval", value: interval }, ...tags];
}

// src/index.common.js
var DEFAULT_GATEWAY_URL = "https://arweave.net";
var DEFAULT_MU_URL = "https://mu.ao-testnet.xyz";
var DEFAULT_CU_URL = "https://cu.ao-testnet.xyz";
var DEFAULT_RELAY_URL = "http://relay.ao-hb.xyz";
var DEFAULT_RELAY_CU_URL = "http://cu.s451-comm3-main.xyz";
var DEFAULT_RELAY_MU_URL = "http://mu.s451-comm3-main.xyz";
var DEFAULT_DEVICE = "relay@1.0";
var defaultFetch = fetch;
function connectWith({ createDataItemSigner: createDataItemSigner3, createSigner: createSigner3 }) {
  const _logger = createLogger();
  function relayMode({
    MODE,
    signer,
    GRAPHQL_URL: GRAPHQL_URL2,
    GRAPHQL_MAX_RETRIES: GRAPHQL_MAX_RETRIES2,
    GRAPHQL_RETRY_BACKOFF: GRAPHQL_RETRY_BACKOFF2,
    MU_URL: MU_URL2 = DEFAULT_RELAY_MU_URL,
    CU_URL: CU_URL2 = DEFAULT_RELAY_CU_URL,
    RELAY_URL: RELAY_URL2 = DEFAULT_RELAY_URL,
    fetch: originalFetch = defaultFetch
  }) {
    const logger = _logger.child("relay");
    logger("Mode Activated \u{1F500}");
    if (!signer) {
      throw new Error("relay mode requires providing a signer to connect()");
    }
    const relayDataItemSigner = signer ? () => signer : createDataItemSigner3;
    const relaySigner = signer ? () => signer : createSigner3;
    const fetch2 = relayerWith({
      /**
       * Always wrap default fetch with relayer,
       * no embellishment beyond the relayer
       */
      fetch: defaultFetch,
      logger,
      HB_URL: RELAY_URL2,
      signer
    });
    const requestLogger = logger.child("request");
    const request = requestWith2({
      logger: requestLogger,
      request: requestWith({
        fetch: defaultFetch,
        logger: requestLogger,
        HB_URL: RELAY_URL2,
        signer
      })
    });
    const { validate } = (0, import_ao_scheduler_utils.connect)({
      cacheSize: 100,
      GRAPHQL_URL: GRAPHQL_URL2,
      GRAPHQL_MAX_RETRIES: GRAPHQL_MAX_RETRIES2,
      GRAPHQL_RETRY_BACKOFF: GRAPHQL_RETRY_BACKOFF2
    });
    const resultLogger = logger.child("result");
    const result2 = resultWith({
      signer,
      loadResult: loadResultWith2({
        fetch: fetch2,
        CU_URL: CU_URL2,
        logger: resultLogger
      }),
      logger: resultLogger
    });
    const messageLogger = logger.child("message");
    const message2 = messageWith({
      signer,
      deployMessage: deployMessageWith2({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: messageLogger
      }),
      logger: messageLogger
    });
    const spawnLogger = logger.child("spawn");
    const spawn2 = spawnWith({
      signer,
      loadTransactionMeta: loadTransactionMetaWith({
        fetch: originalFetch,
        GRAPHQL_URL: GRAPHQL_URL2,
        logger: spawnLogger
      }),
      validateScheduler: validate,
      deployProcess: deployProcessWith2({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: spawnLogger
      }),
      logger: spawnLogger
    });
    const monitorLogger = logger.child("monitor");
    const monitor2 = monitorWith({
      signer,
      deployMonitor: deployMonitorWith({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: monitorLogger
      }),
      logger: monitorLogger
    });
    const unmonitorLogger = logger.child("unmonitor");
    const unmonitor2 = unmonitorWith({
      signer,
      deployUnmonitor: deployUnmonitorWith({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: unmonitorLogger
      }),
      logger: monitorLogger
    });
    const resultsLogger = logger.child("results");
    const results2 = resultsWith({
      signer,
      queryResults: queryResultsWith2({
        fetch: fetch2,
        CU_URL: CU_URL2,
        logger: resultsLogger
      }),
      logger: resultsLogger
    });
    const dryrunLogger = logger.child("dryrun");
    const dryrun2 = dryrunWith({
      signer,
      dryrunFetch: dryrunFetchWith({
        fetch: fetch2,
        CU_URL: CU_URL2,
        logger: dryrunLogger
      }),
      logger: dryrunLogger
    });
    const assignLogger = logger.child("assign");
    const assign2 = assignWith({
      signer,
      deployAssign: deployAssignWith({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: assignLogger
      }),
      logger: messageLogger
    });
    return {
      MODE,
      request,
      result: result2,
      results: results2,
      message: message2,
      spawn: spawn2,
      monitor: monitor2,
      unmonitor: unmonitor2,
      dryrun: dryrun2,
      assign: assign2,
      ccreateDataItemSigner: relayDataItemSigner,
      createSigner: relaySigner
    };
  }
  function legacyMode({
    MODE,
    GRAPHQL_URL: GRAPHQL_URL2,
    GRAPHQL_MAX_RETRIES: GRAPHQL_MAX_RETRIES2,
    GRAPHQL_RETRY_BACKOFF: GRAPHQL_RETRY_BACKOFF2,
    MU_URL: MU_URL2 = DEFAULT_MU_URL,
    CU_URL: CU_URL2 = DEFAULT_CU_URL,
    fetch: fetch2 = defaultFetch,
    noLog
  }) {
    const logger = _logger.child("legacy");
    if (!noLog) logger("Mode Activated \u2139\uFE0F");
    const { validate } = (0, import_ao_scheduler_utils.connect)({
      cacheSize: 100,
      GRAPHQL_URL: GRAPHQL_URL2,
      GRAPHQL_MAX_RETRIES: GRAPHQL_MAX_RETRIES2,
      GRAPHQL_RETRY_BACKOFF: GRAPHQL_RETRY_BACKOFF2
    });
    const resultLogger = logger.child("result");
    const result2 = resultWith({
      loadResult: loadResultWith2({
        fetch: fetch2,
        CU_URL: CU_URL2,
        logger: resultLogger
      }),
      logger: resultLogger
    });
    const messageLogger = logger.child("message");
    const message2 = messageWith({
      deployMessage: deployMessageWith2({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: messageLogger
      }),
      logger: messageLogger
    });
    const spawnLogger = logger.child("spawn");
    const spawn2 = spawnWith({
      loadTransactionMeta: loadTransactionMetaWith({
        fetch: fetch2,
        GRAPHQL_URL: GRAPHQL_URL2,
        logger: spawnLogger
      }),
      validateScheduler: validate,
      deployProcess: deployProcessWith2({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: spawnLogger
      }),
      logger: spawnLogger
    });
    const monitorLogger = logger.child("monitor");
    const monitor2 = monitorWith({
      deployMonitor: deployMonitorWith({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: monitorLogger
      }),
      logger: monitorLogger
    });
    const unmonitorLogger = logger.child("unmonitor");
    const unmonitor2 = unmonitorWith({
      deployUnmonitor: deployUnmonitorWith({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: unmonitorLogger
      }),
      logger: monitorLogger
    });
    const resultsLogger = logger.child("results");
    const results2 = resultsWith({
      queryResults: queryResultsWith2({
        fetch: fetch2,
        CU_URL: CU_URL2,
        logger: resultsLogger
      }),
      logger: resultsLogger
    });
    const dryrunLogger = logger.child("dryrun");
    const dryrun2 = dryrunWith({
      dryrunFetch: dryrunFetchWith({
        fetch: fetch2,
        CU_URL: CU_URL2,
        logger: dryrunLogger
      }),
      logger: dryrunLogger
    });
    const assignLogger = logger.child("assign");
    const assign2 = assignWith({
      deployAssign: deployAssignWith({
        fetch: fetch2,
        MU_URL: MU_URL2,
        logger: assignLogger
      }),
      logger: messageLogger
    });
    return {
      MODE,
      result: result2,
      results: results2,
      message: message2,
      spawn: spawn2,
      monitor: monitor2,
      unmonitor: unmonitor2,
      dryrun: dryrun2,
      assign: assign2,
      createDataItemSigner: createDataItemSigner3
    };
  }
  function mainnetMode({
    MODE,
    signer,
    GRAPHQL_URL: GRAPHQL_URL2,
    device = DEFAULT_DEVICE,
    URL: URL2 = DEFAULT_RELAY_URL,
    MU_URL: MU_URL2 = DEFAULT_RELAY_MU_URL,
    CU_URL: CU_URL2 = DEFAULT_RELAY_CU_URL,
    fetch: fetch2 = defaultFetch
  }) {
    const isRelayMode = device === "relay@1.0";
    const logger = isRelayMode ? _logger.child("mainnet-relay") : _logger.child("mainnet-process");
    logger("Mode Activated %s", isRelayMode ? "\u{1F500}" : "\u{1F432}");
    if (!signer) {
      throw new Error("mainnet mode requires providing a signer to connect()");
    }
    const mainnetDataItemSigner = signer ? () => signer : createDataItemSigner3;
    const mainnetSigner = signer ? () => signer : createSigner3;
    const relayFetch = relayerWith({
      /**
       * Always wrap default fetch with relayer,
       * no embellishment beyond the relayer
       */
      fetch: defaultFetch,
      logger,
      HB_URL: URL2,
      signer
    });
    async function mockValidate(address) {
      logger('Mock validation for address "%s"', address);
      return true;
    }
    const resultLogger = logger.child("result");
    const loadResult = isRelayMode ? loadResultWith2({
      fetch: relayFetch,
      logger: resultLogger,
      HB_URL: URL2,
      CU_URL: CU_URL2,
      signer
    }) : loadResultWith({
      fetch: defaultFetch,
      logger: resultLogger,
      HB_URL: URL2,
      signer
    });
    const result2 = resultWith({
      signer,
      loadResult,
      logger: resultLogger
    });
    const messageLogger = logger.child("message");
    const deployMessage = isRelayMode ? deployMessageWith2({
      fetch: isRelayMode ? relayFetch : defaultFetch,
      logger: messageLogger,
      HB_URL: URL2,
      MU_URL: MU_URL2,
      CU_URL: CU_URL2,
      signer
    }) : deployMessageWith({
      fetch: defaultFetch,
      logger: messageLogger,
      HB_URL: URL2,
      signer
    });
    const message2 = messageWith({
      signer,
      deployMessage,
      logger: messageLogger
    });
    const resultsLogger = logger.child("results");
    const queryResults = isRelayMode ? queryResultsWith2({
      fetch: relayFetch,
      CU_URL: CU_URL2,
      logger: resultsLogger
    }) : queryResultsWith({
      fetch: defaultFetch,
      HB_URL: URL2,
      logger: resultsLogger,
      signer
    });
    const getMessageById2 = messageIdWith({
      getMessageId: getMessageById({ fetch: fetch2, locate: import_ao_scheduler_utils.locate })
    });
    const getMessages = messagesWith({
      messages: getMessagesByRange({ fetch: fetch2, locate: import_ao_scheduler_utils.locate })
    });
    const getLastSlot = processIdWith({
      processId: getLastSlotWith({
        fetch: fetch2,
        locate: import_ao_scheduler_utils.locate
      })
    });
    const results2 = resultsWith({
      signer,
      queryResults,
      logger: resultsLogger
    });
    const spawnLogger = logger.child("spawn");
    const deployProcess = isRelayMode ? deployProcessWith2({
      fetch: relayFetch,
      HB_URL: URL2,
      MU_URL: MU_URL2,
      logger: spawnLogger
    }) : deployProcessWith({
      fetch: defaultFetch,
      logger: spawnLogger,
      HB_URL: URL2,
      signer
    });
    const spawn2 = spawnWith({
      signer,
      loadTransactionMeta: loadTransactionMetaWith({
        fetch: fetch2,
        GRAPHQL_URL: GRAPHQL_URL2,
        logger: spawnLogger
      }),
      validateScheduler: mockValidate,
      deployProcess,
      logger: spawnLogger
    });
    const dryrunLogger = logger.child("dryrun");
    const dryrun2 = dryrunWith({
      signer,
      dryrunFetch: dryrunFetchWith({
        fetch: isRelayMode ? relayFetch : fetch2,
        CU_URL: CU_URL2,
        logger: dryrunLogger
      }),
      logger: dryrunLogger
    });
    const requestLogger = logger.child("request");
    const request = requestWith2({
      signer,
      logger: requestLogger,
      MODE,
      method: "GET",
      device,
      dryrun: dryrun2,
      message: message2,
      result: result2,
      spawn: spawn2,
      request: requestWith({
        fetch: defaultFetch,
        logger: requestLogger,
        HB_URL: URL2,
        signer
      })
    });
    const getLogger = logger.child("get");
    const get = requestWith2({
      logger: getLogger,
      MODE,
      method: "GET",
      device,
      dryrun: dryrun2,
      message: message2,
      result: result2,
      spawn: spawn2,
      signer,
      request: requestWith({
        fetch: defaultFetch,
        method: "GET",
        logger: getLogger,
        HB_URL: URL2,
        signer
      })
    });
    const postLogger = logger.child("post");
    const post = requestWith2({
      signer,
      logger: postLogger,
      MODE,
      method: "POST",
      device,
      dryrun: dryrun2,
      message: message2,
      result: result2,
      spawn: spawn2,
      request: requestWith({
        fetch: defaultFetch,
        method: "GET",
        logger: postLogger,
        HB_URL: URL2,
        signer
      })
    });
    return {
      MODE: isRelayMode ? "relay" : "mainnet",
      request,
      getMessages,
      getLastSlot,
      getMessageById: getMessageById2,
      get,
      post,
      result: result2,
      message: message2,
      spawn: spawn2,
      results: results2,
      getOperator: getOperator({ fetch: fetch2, URL: URL2 }),
      getNodeBalance: getNodeBalance({
        request: requestWith({
          fetch: defaultFetch,
          method: "GET",
          logger: postLogger,
          HB_URL: URL2,
          signer
        })
      }),
      createDataItemSigner: mainnetDataItemSigner,
      createSigner: mainnetSigner
    };
  }
  function connect2(args = {}) {
    let { GRAPHQL_URL: GRAPHQL_URL2, GATEWAY_URL: GATEWAY_URL2 = DEFAULT_GATEWAY_URL, ...restArgs } = args;
    if (!GRAPHQL_URL2) {
      GRAPHQL_URL2 = joinUrl({ url: GATEWAY_URL2, path: "/graphql" });
    }
    const MODE = args.MODE || "legacy";
    if (MODE === "legacy") return legacyMode({ ...restArgs, GRAPHQL_URL: GRAPHQL_URL2 });
    if (MODE === "mainnet") return mainnetMode({ ...restArgs, GRAPHQL_URL: GRAPHQL_URL2 });
    throw new Error(`Unrecognized MODE: ${MODE}`);
  }
  return connect2;
}

// src/client/node/wallet.js
var wallet_exports = {};
__export(wallet_exports, {
  createDataItemSigner: () => createDataItemSigner,
  createSigner: () => createSigner
});
var import_node_crypto = require("node:crypto");
function createANS104Signer({ privateKey, publicKey, address }) {
  const signer = async (create) => {
    const deepHash = await create({
      type: 1,
      publicKey,
      alg: "rsa-v1_5-sha256"
    });
    const signature = (0, import_node_crypto.createSign)("sha256").update(deepHash).sign({ key: privateKey, padding: import_node_crypto.constants.RSA_PKCS1_PSS_PADDING });
    return { signature, address };
  };
  return signer;
}
function createHttpSigner({ publicKey, privateKey, address }) {
  const signer = async (create) => {
    const signatureBase = await create({
      type: 1,
      publicKey,
      alg: "rsa-pss-sha512"
    });
    const signature = (0, import_node_crypto.createSign)("sha512").update(signatureBase).sign({ key: privateKey, padding: import_node_crypto.constants.RSA_PKCS1_PSS_PADDING });
    return { signature, address };
  };
  return signer;
}
function createSigner(wallet) {
  const publicKey = Buffer.from(wallet.n, "base64url");
  const privateKey = (0, import_node_crypto.createPrivateKey)({ key: wallet, format: "jwk" });
  const address = (0, import_node_crypto.createHash)("sha256").update(publicKey).digest("base64url");
  const dataItemSigner = createANS104Signer({ wallet, privateKey, publicKey, address });
  const httpSigner = createHttpSigner({ wallet, publicKey, privateKey, address });
  const signer = (create, kind) => {
    if (kind === DATAITEM_SIGNER_KIND) return dataItemSigner(create);
    if (kind === HTTP_SIGNER_KIND) return httpSigner(create);
    throw new Error(`signer kind unknown "${kind}"`);
  };
  return signer;
}
var createDataItemSigner = createSigner;

// src/index.js
var GATEWAY_URL = process.env.GATEWAY_URL || void 0;
var MU_URL = process.env.MU_URL || void 0;
var CU_URL = process.env.CU_URL || void 0;
var GRAPHQL_URL = process.env.GRAPHQL_URL || void 0;
var GRAPHQL_MAX_RETRIES = process.env.GRAPHQL_MAX_RETRIES || void 0;
var GRAPHQL_RETRY_BACKOFF = process.env.GRAPHQL_RETRY_BACKOFF || void 0;
var RELAY_URL = process.env.RELAY_URL || void 0;
var AO_URL = process.env.AO_URL = void 0;
var connect = connectWith({
  createDataItemSigner: wallet_exports.createDataItemSigner,
  createSigner: wallet_exports.createSigner
});
var createDataItemSigner2 = wallet_exports.createDataItemSigner;
var createSigner2 = wallet_exports.createSigner;
var { result, results, message, spawn, monitor, unmonitor, dryrun, assign } = connect({
  MODE: "legacy",
  GATEWAY_URL,
  MU_URL,
  CU_URL,
  RELAY_URL,
  AO_URL,
  GRAPHQL_URL,
  GRAPHQL_MAX_RETRIES,
  GRAPHQL_RETRY_BACKOFF,
  noLog: true
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  assign,
  connect,
  createDataItemSigner,
  createSigner,
  dryrun,
  message,
  monitor,
  result,
  results,
  serializeCron,
  spawn,
  unmonitor
});
